File: algorithms/cpp/_023_MergekSortedLists/_023_MergekSortedLists_std.cpp
#include <iostream>
#include <vector>

using namespace std;


//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;

    explicit ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    /**
     * @param lists: a list of ListNode
     * @return: The head of one sorted list.
     */
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        // write your code here
        if (lists.size() == 0) return nullptr;
        if (lists.size() == 1) return lists[0];
        ListNode *merged = mergeHelper(lists, 0, lists.size() - 1);
        return merged;
    }

    ListNode *mergeHelper(vector<ListNode *> &lists, int start, int end) {
        if (start >= end) return lists[start];
        int mid = (end - start) / 2 + start;
        ListNode *left = mergeHelper(lists, start, mid);
        ListNode *right = mergeHelper(lists, mid + 1, end);
        return merge(left, right);
    }

    ListNode *merge(ListNode *a, ListNode *b) {
        ListNode *dummy = new ListNode(0);
        ListNode *cur = dummy;
        while (a && b) {
            if (a->val < b->val) {
                cur->next = a;
                cur = cur->next;
                a = a->next;
            } else {
                cur->next = b;
                cur = cur->next;
                b = b->next;
            }
        }
        while (a) {
            cur->next = a;
            cur = cur->next;
            a = a->next;
        }
        while (b) {
            cur->next = b;
            cur = cur->next;
            b = b->next;
        }
        return dummy->next;
    }
};

int main() {
    auto *so = new Solution();
    cout << "new file!" << endl;
    return 0;
}

File: algorithms/java/src/_023_MergekSortedLists/_023_MergekSortedLists_std.java
package _023_MergekSortedLists;


import java.util.List;

public class _023_MergekSortedLists_std {
    public static void main(String[] args) {
        _023_MergekSortedLists_std_ so = new _023_MergekSortedLists_std_();

    }
}

class _023_MergekSortedLists_std_ {

}


//Definition for ListNode.
class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
        this.next = null;
    }
}

class Solution {
    /**
     * @param lists: a list of ListNode
     * @return: The head of one sorted list.
     */
    public ListNode mergeKLists(List<ListNode> lists) {
        // write your code here
        return helper(lists, 0, lists.size() - 1);
    }

    private ListNode helper(List<ListNode> lists, int start, int end) {
        if (start > end) return null;
        if (start == end) return lists.get(start);
        int mid = (end - start) / 2 + start;
        ListNode left = helper(lists, start, mid);
        ListNode right = helper(lists, mid + 1, end);
        return mergeTwoLists(left, right);
    }

    private ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        if (l2 == null) return l1;
        ListNode head = new ListNode(0);
        ListNode p = head;
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                p.next = l1;
                l1 = l1.next;
            } else {
                p.next = l2;
                l2 = l2.next;
            }
            p = p.next;
        }

        if (l1 != null) p.next = l1;
        if (l2 != null) p.next = l2;
        return head.next;
    }
}